File: src/types/character.ts
export type CharacterSetupStage =
  | 'idle'
  | 'detecting'
  | 'multiple_detected'
  | 'ready_to_generate'
  | 'generating'
  | 'generated'
  | 'error'

export type CharacterSetupErrorCode =
  | 'invalid_file'
  | 'no_person'
  | 'detection_failed'
  | 'backend_unavailable'
  | 'generation_failed'
  | 'base_image_download_failed'
  | 'transparent_asset_failed'
  | 'normalization_failed'
  | 'asset_write_failed'
  | 'master_character_not_found'
  | 'master_character_invalid'
  | 'invalid_pet_state'
  | 'network_failed'
  | 'request_timeout'

export interface CharacterSetupError {
  code: CharacterSetupErrorCode
  message: string
  retryable: boolean
}

export interface DetectedPersonBox {
  id: string
  x: number
  y: number
  width: number
  height: number
  score: number
}

export type PetVisualState =
  | 'idle'
  | 'walking'
  | 'thinking'
  | 'working'
  | 'waiting'
  | 'celebrating'

export type AlphaBoundingBox = {
  left: number
  top: number
  right: number
  bottom: number
}

export type CharacterVisualMetrics = {
  sourceWidth: number
  sourceHeight: number
  subjectHeightRatio: number
  baselineRatio: number
  measurement: 'alpha-bbox-v1'
  alphaThreshold: number
  alphaBBox?: AlphaBoundingBox
  aggregation?: 'single' | 'median'
  sampleCount?: number
  subjectHeightRatioRange?: [number, number]
  baselineRatioRange?: [number, number]
}

export interface CharacterStateAsset {
  assetId: string
  state: PetVisualState
  baseImage: string
  transparentImage: string
  normalizedImage: string
  modelName: string
  promptVersion: string
  createdAt: string
  providerHttpStatus: number
  durationMs: number
  visualMetrics?: CharacterVisualMetrics
}

export type CharacterStateAssets = Partial<Record<PetVisualState, CharacterStateAsset>>

export type CharacterMotionFrame = {
  frameIndex: number
  sourceFrameIndex?: number
  imageUrl: string
  format: 'png' | 'webp'
  width: number
  height: number
  visualMetrics?: CharacterVisualMetrics
}

export type WalkingSourceMetadata = {
  provider: string
  modelName: string
  modelId?: string
  sourceVideoUrl: string
  sourceVideoDurationSec: number
  sourceVideoFps: number
  sourceVideoWidth: number
  sourceVideoHeight: number
  cycleStartFrame?: number
  cycleEndFrame?: number
  taskId?: string
  callIndex?: number
}

export type WalkingMotionAsset = {
  version: string
  status: 'not_started' | 'generating' | 'processing' | 'partial' | 'completed' | 'error'
  frames: CharacterMotionFrame[]
  frameCount: number
  playbackFps: number
  frameDurationMs: number
  source: WalkingSourceMetadata
  promptVersion?: string
  errorMessage?: string
  createdAt: string
  updatedAt: string
  visualMetrics?: CharacterVisualMetrics
}

export type CharacterMotionAssets = {
  walking?: WalkingMotionAsset
}

export interface CharacterGenerationResult {
  characterId: string
  sourceImage: string
  croppedPersonImage: string
  baseImage: string
  transparentImage?: string
  normalizedImage?: string
  modelName?: string
  promptVersion?: string
  createdAt: string | number
  stateAssets?: CharacterStateAssets
  motionAssets?: CharacterMotionAssets
  visualMetrics?: CharacterVisualMetrics
}

export interface CharacterSetupSession {
  stage: CharacterSetupStage
  fileName: string
  sourcePhoto?: string
  people: DetectedPersonBox[]
  selectedPersonId?: string
  croppedPersonImage?: string
  result?: CharacterGenerationResult
  error?: CharacterSetupError
  petName: string
  personality: string
  motionStyle: string
}

export interface CharacterGenerationRequest {
  imageBase64: string
  petName: string
  personality: string
  motionStyle: string
}

export interface CharacterGenerationResponse {
  success: true
  characterId: string
  baseImage: string
  transparentImage: string
  normalizedImage: string
  modelName: string
  promptVersion: string
  createdAt: string
}

export interface CharacterStateAssetRequest {
  characterId: string
  state: PetVisualState
}

export interface CharacterStateAssetResponse {
  success: true
  characterId: string
  masterImage: string
  asset: CharacterStateAsset
}

File: server/schemas/character.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator


PetVisualState = Literal[
    "idle",
    "walking",
    "thinking",
    "working",
    "waiting",
    "celebrating",
]


class CharacterGenerationRequest(BaseModel):
    imageBase64: str = Field(min_length=32, max_length=16_000_000)
    petName: str = Field(default="桌宠", min_length=1, max_length=24)
    personality: str = Field(default="friendly", max_length=32)
    motionStyle: str = Field(default="light", max_length=32)

    @field_validator("imageBase64")
    @classmethod
    def validate_image_data_url(cls, value: str) -> str:
        supported_prefixes = (
            "data:image/jpeg;base64,",
            "data:image/png;base64,",
            "data:image/webp;base64,",
        )
        if not value.startswith(supported_prefixes):
            raise ValueError("imageBase64 must be a JPEG, PNG or WebP data URL")
        return value


class CharacterGenerationResponse(BaseModel):
    success: bool = True
    characterId: str
    baseImage: str
    transparentImage: str
    normalizedImage: str
    modelName: str
    promptVersion: str
    createdAt: str


class CharacterStateAssetRequest(BaseModel):
    characterId: str = Field(pattern=r"^char_[A-Za-z0-9_-]{4,64}$")
    state: PetVisualState


class CharacterStateAsset(BaseModel):
    assetId: str
    state: PetVisualState
    baseImage: str
    transparentImage: str
    normalizedImage: str
    modelName: str
    promptVersion: str
    createdAt: str
    providerHttpStatus: int
    durationMs: int


class CharacterStateAssetResponse(BaseModel):
    success: bool = True
    characterId: str
    masterImage: str
    asset: CharacterStateAsset


class WalkingMotionRequest(BaseModel):
    characterId: str = Field(pattern=r"^char_[A-Za-z0-9_-]{4,64}$")


class WalkingMotionFrame(BaseModel):
    frameIndex: int
    sourceFrameIndex: Optional[int] = None
    imageUrl: str
    format: Literal["png", "webp"]
    width: int
    height: int


class WalkingSourceMetadata(BaseModel):
    provider: str
    modelName: str
    modelId: Optional[str] = None
    sourceVideoUrl: str
    sourceVideoDurationSec: float
    sourceVideoFps: float
    sourceVideoWidth: int
    sourceVideoHeight: int
    cycleStartFrame: Optional[int] = None
    cycleEndFrame: Optional[int] = None
    taskId: Optional[str] = None
    callIndex: Optional[int] = None


class WalkingMotionAsset(BaseModel):
    version: str
    status: Literal["not_started", "generating", "processing", "partial", "completed", "error"]
    frames: list[WalkingMotionFrame]
    frameCount: int
    playbackFps: float
    frameDurationMs: float
    source: WalkingSourceMetadata
    promptVersion: Optional[str] = None
    errorMessage: Optional[str] = None
    createdAt: str
    updatedAt: str


class WalkingMotionResponse(BaseModel):
    success: bool = True
    characterId: str
    motion: WalkingMotionAsset


class HealthResponse(BaseModel):
    status: str
    provider: str
    modelConfigured: bool
    apiKeyConfigured: bool
    deepseekConfigured: bool
